### PR TITLE
New 'skipped' field in semgrep-core json response

### PR DESCRIPTION
**Pull request by mjambon**
This is the ocaml side of https://github.com/returntocorp/semgrep/issues/3712.

The new field `skipped` is a list of targets that were "skipped", which loosely means it didn't go through the usual semgrep parsing+matching and is worth reporting. 

I moved the type definition out of spacegrep to its own library (module `Semgrep_core_response_t`). Everything in semgrep can depend on it.

Here's what I tested:

Binary files skipped by spacegrep:
```
~/semgrep $ spacegrep --output-format semgrep foo semgrep-core/src/spacegrep/perf-data/django-template.po.gz | ydump
{
  "matches": [],
  "errors": [],
  "skipped": [
    {
      "path": "semgrep-core/src/spacegrep/perf-data/django-template.po.gz",
      "reason": "binary",
      "details": "target looks like a binary file"
    }
  ],
  "stats": { "okfiles": 0, "errorfiles": 0 }
}
```

Minified files skipped by spacegrep:
```
~/semgrep $ spacegrep --output-format semgrep foo ./semgrep-core/src/pfff/tests/js/parsing/big/minimized.min.js | ydump
{
  "matches": [],
  "errors": [],
  "skipped": [
    {
      "path": "./semgrep-core/src/pfff/tests/js/parsing/big/minimized.min.js",
      "reason": "minified",
      "details": "not a source file: target file appears to be minified"
    }
  ],
  "stats": { "okfiles": 0, "errorfiles": 0 }
}
```

Too many matches by semgrep-core:
```
~/semgrep $ semgrep-core -lang ml -json -e '$X' semgrep-core/src/cli -max_match_per_file 10 | ydump
{
  "matches": [
    {
      "check_id": "-e/-f",
      "path": "semgrep-core/src/cli/version.ml",
      "start": { "line": 6, "col": 15, "offset": 101 },
      "end": { "line": 6, "col": 37, "offset": 123 },
      "extra": {
        "message": "",
        "metavars": {
          "$X": {
            "start": { "line": 6, "col": 15, "offset": 101 },
            "end": { "line": 6, "col": 37, "offset": 123 },
            "abstract_content": "\"v0.62.0-12-g123bd463\"",
            "unique_id": {
              "type": "AST",
              "md5sum": "0f0cd06cb9585da45d207bad8903df17"
            }
          }
        },
        "lines": []
      }
    },
    {
      "check_id": "-e/-f",
      "path": "semgrep-core/src/cli/version.ml",
      "start": { "line": 6, "col": 5, "offset": 91 },
      "end": { "line": 6, "col": 37, "offset": 123 },
      "extra": {
        "message": "",
        "metavars": {
          "$X": {
            "start": { "line": 6, "col": 5, "offset": 91 },
            "end": { "line": 6, "col": 37, "offset": 123 },
            "abstract_content": "version\"v0.62.0-12-g123bd463\"",
            "unique_id": {
              "type": "AST",
              "md5sum": "ba3fe90d74a320900d18dfce5d8999c6"
            }
          }
        },
        "lines": []
      }
    },
    {
      "check_id": "-e/-f",
      "path": "semgrep-core/src/cli/version.ml",
      "start": { "line": 6, "col": 5, "offset": 91 },
      "end": { "line": 6, "col": 12, "offset": 98 },
      "extra": {
        "message": "",
        "metavars": {
          "$X": {
            "start": { "line": 6, "col": 5, "offset": 91 },
            "end": { "line": 6, "col": 12, "offset": 98 },
            "abstract_content": "version",
            "unique_id": { "type": "id", "sid": 1 }
          }
        },
        "lines": []
      }
    }
  ],
  "errors": [
    {
      "check_id": "TooManyMatches",
      "path": "semgrep-core/src/cli/Experiments.ml",
      "start": { "line": 1, "col": 1, "offset": 0 },
      "end": { "line": 1, "col": 1, "offset": 0 },
      "extra": { "message": "too many matches: $X", "line": "open Common" }
    },
    {
      "check_id": "TooManyMatches",
      "path": "semgrep-core/src/cli/Main.ml",
      "start": { "line": 1, "col": 1, "offset": 0 },
      "end": { "line": 1, "col": 1, "offset": 0 },
      "extra": {
        "message": "too many matches: $X",
        "line": "(*s: semgrep/CLI/Main.ml *)"
      }
    }
  ],
  "skipped": [
    {
      "path": "semgrep-core/src/cli/Experiments.ml",
      "reason": "too_many_matches",
      "details":
        "found 494 matches for rule -e/-f, which exceeds the maximum of 10 matches.",
      "skipped_rule": "-e/-f"
    },
    {
      "path": "semgrep-core/src/cli/Main.ml",
      "reason": "too_many_matches",
      "details":
        "found 2970 matches for rule -e/-f, which exceeds the maximum of 10 matches.",
      "skipped_rule": "-e/-f"
    }
  ],
  "stats": { "okfiles": 1, "errorfiles": 2 }
}
```

I'm giving up on trying to trigger `Irrelevant_rules`.

PR checklist:
- [x] documentation is up to date
- [x] changelog is up to date

### CODE REVIEW

**Comment by IagoAbal** [comment]
Irrelevant rules should be the easiest one?

**Review by IagoAbal — APPROVED** [review]

**Review by IagoAbal — CHANGES_REQUESTED** [review]
Maybe you should take a look at the irrelevant-rule part, I will try take a look myself too.

This rule:
```
rules:
- id: test
  languages: [python]
  patterns:
  - pattern: abc
  message: Test
  severity: ERROR
```

On this file:
```
x += 1
```

Passing `-filter_irrelevant_rules` will skip the file, and logging seems to confirm that, but I don't see that reported in the JSON output.

File: semgrep-core/src/cli/Main.ml [hunk 1]
```diff
@@ -729,17 +754,24 @@ let semgrep_with_patterns lang (rules, rule_parse_time) files_or_dirs =
...
     (List.length res.RP.matches)
     (List.length res.RP.errors);
-  let matches, new_errors =
+  let matches, new_errors, skipped =
     filter_files_with_too_many_matches_and_transform_as_timeout res.RP.matches
   in
   let errors = new_errors @ res.RP.errors in
-  let res = { RP.matches; errors; rule_profiling = res.RP.rule_profiling } in
+  let res =
+    { RP.matches; errors; skipped; rule_profiling = res.RP.rule_profiling }
```
Comment (IagoAbal) [inline semgrep-core/src/cli/Main.ml:774]: Here you are just taking the `skipped` returned by `filter_files_with_too_many_matches_and_transform_as_timeout` but you are throwing away `res.skipped` which contains those skipped by `-filter_irrelevant_rules`.

Suggestion:
```
  let matches, new_errors, new_skipped =
     filter_files_with_too_many_matches_and_transform_as_timeout res.RP.matches
   in
 let errors = new_errors @ res.RP.errors in
 let skipped = new_skipped @ res.RP.skipped in
 let res =
     { RP.matches; errors; skipped; rule_profiling = res.RP.rule_profiling }
```

Comment (mjambon) [reply semgrep-core/src/cli/Main.ml:774]: wow, well-spotted!

File: semgrep-core/src/cli/Main.ml [hunk 2]
```diff
@@ -834,11 +867,13 @@ let semgrep_with_rules (rules, rule_parse_time) files_or_dirs =
   let res = RP.make_rule_result file_results !report_time rule_parse_time in
   logger#info "found %d matches and %d errors" (List.length res.matches)
     (List.length res.errors);
-  let matches, new_errors =
+  let matches, new_errors, skipped =
```
Comment (IagoAbal) [inline semgrep-core/src/cli/Main.ml]: I didn't realize we had two occurrences of this code. We have the same problem here.

Comment (mjambon) [reply semgrep-core/src/cli/Main.ml]: :+1:

**Comment by mjambon** [comment]
:tada: 
```
$ semgrep-core -lang py -config abc.yaml -filter_irrelevant_rules abc.py -json | ydump
{
  "matches": [],
  "errors": [],
  "skipped": [
    {
      "path": "abc.py",
      "reason": "irrelevant_rule",
      "details": "target doesn't contain some elements required by the rule",
      "skipped_rule": "test"
    }
  ],
  "stats": { "okfiles": 1, "errorfiles": 0 }
}
```

**Pull request merged**